internal/types/compare.go

```diff
@@ -40,12 +40,67 @@ const (
 )
 
 // Compare compares any BSON values in the same way as MongoDB does it for filtering.
+// When it compares scalar filterValue with array docValue, it returns first item
+// which is not Incomparable.
 //
 // It converts types as needed; that may result in different types being equal.
 // For that reason, it typically should not be used in tests.
 //
 // Compare and contrast with test helpers in testutil package.
 func Compare(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{greater: true, less: true, equal: true})
+}
+
+// CompareGreaterThan compares docValue for the greater than condition.
+func CompareGreaterThan(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{greater: true})
+}
+
+// CompareGreaterThanOrEq compares docValue for the greater than or equal to condition.
+func CompareGreaterThanOrEq(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{greater: true, equal: true})
+}
+
+// CompareLessThan compares docValue for the less than condition.
+func CompareLessThan(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{less: true})
+}
+
+// CompareLessThanOrEq compares docValue for the less than or equal to condition.
+func CompareLessThanOrEq(docValue, filterValue any) CompareResult {
+	return compare(docValue, filterValue, cond{less: true, equal: true})
+}
+
+// cond is the condition used to decide whether to return the
+// comparison result or keep iterating on the array docValue.
+type cond struct {
+	greater bool
+	less    bool
+	equal   bool
+}
+
+// satisfies checks if the result satisfies the condition.
+func (c cond) satisfies(result CompareResult) bool {
+	switch result {
+	case Greater:
+		return c.greater
+	case Less:
+		return c.less
+	case Equal:
+		return c.equal
+	case Incomparable:
+		return false
+	default:
+		return false
```

Comment:
What could be other values apart from Greater, Less, Equal and Incomparable?

Reply:
Removed this default added panic, just for the consistency with rest of the code